File: packages/ruby/lib/spikard/service_http_additions.rb
# frozen_string_literal: true

def on_request(&block)
  # Called before any other processing for each inbound request.
  @lifecycle_hooks.push(["on_request", block])
  self
end

def pre_validation(&block)
  # Called after parsing but before parameter validation.
  @lifecycle_hooks.push(["pre_validation", block])
  self
end

def pre_handler(&block)
  # Called after validation but before invoking the route handler.
  @lifecycle_hooks.push(["pre_handler", block])
  self
end

def on_response(&block)
  # Called after the handler returns but before the response is serialized.
  @lifecycle_hooks.push(["on_response", block])
  self
end

def on_error(&block)
  # Called when a handler returns an error.
  @lifecycle_hooks.push(["on_error", block])
  self
end

def websocket(path, &block)
  # Register a WebSocket upgrade handler at the given path.
  @registrations.push(["websocket", [path], block])
  self
end

def sse(path, &block)
  # Register an SSE event producer at the given path.
  @registrations.push(["sse", [path], block])
  self
end

File: crates/spikard-node/service_http_additions.ts
  /**
   * Called before any other processing for each inbound request.
   */
  onRequest(handler: (...args: any[]) => Promise<void>): this {
    this._app.onRequest(handler);
    return this;
  }
  /**
   * Called after parsing but before parameter validation.
   */
  preValidation(handler: (...args: any[]) => Promise<void>): this {
    this._app.preValidation(handler);
    return this;
  }
  /**
   * Called after validation but before invoking the route handler.
   */
  preHandler(handler: (...args: any[]) => Promise<void>): this {
    this._app.preHandler(handler);
    return this;
  }
  /**
   * Called after the handler returns but before the response is serialized.
   */
  onResponse(handler: (...args: any[]) => Promise<void>): this {
    this._app.onResponse(handler);
    return this;
  }
  /**
   * Called when a handler returns an error.
   */
  onError(handler: (...args: any[]) => Promise<void>): this {
    this._app.onError(handler);
    return this;
  }
